python/lib/client/dmod/client/request_clients.py

```diff
@@ -1,600 +1,823 @@
-                                           data_location=item_name)
-        self.last_response: DatasetManagementResponse = await self.async_make_request(request)
+            reason = f'Unable to Create Parent Directory'
+            msg = f'{self.__class__.__name__} could not create parent directory for downloading item to {str(dest)}'
+            return DatasetManagementResponse(success=False, reason=reason, message=msg)
+
+        try:
+            request, final_response_type = self._request_prep(dataset_name=dataset_name, item_name=item_name,
+                                                              action=ManagementAction.REQUEST_DATA)
+        # TODO: (later) implement and use DmodAuthenticationFailure
+        except DmodRuntimeError as e:
+            reason = f'{self.__class__.__name__} Download Auth Failure'
+            return MaaSDatasetManagementResponse(success=False, reason=reason, message=str(e))
+
         with dest.open('w') as file:
-            for page in range(1, (self.last_response.total_pages + 1)):
-                request = DatasetManagementMessage(action=ManagementAction.DOWNLOAD_DATA, dataset_name=dataset_name,
-                                                   data_location=item_name, page=page)
-                self.last_response: DatasetManagementResponse = await self.async_make_request(request)
-                file.write(self.last_response.file_data)
+            # Do initial request outside of generator
+            await self._transport_client.async_send(data=str(request))
+            async for received_data_msg in self._transfer_receiver():
+                file.write(received_data_msg.data)
+            final_data = await self._transport_client.async_recv()
 
-    async def list_datasets(self, category: Optional[DataCategory] = None) -> List[str]:
-        action = ManagementAction.LIST_ALL if category is None else ManagementAction.SEARCH
-        request = DatasetManagementMessage(action=action, category=category)
-        self.last_response = await self.async_make_request(request)
-        return self._parse_list_of_dataset_names_from_response(self.last_response)
+        try:
+            final_response_json = json.loads(final_data)
+        except Exception as e:
+            msg = f"{self.__class__.__name__} failed with {e.__class__.__name__} parsing `{final_data}` to JSON)"
+            return final_response_type(success=False, reason=f"JSON Parse Failure On Final Response", message=msg)
+
+        final_response = final_response_type.factory_init_from_deserialized_json(final_response_json)
+        if final_response is None:
+            return final_response_type(success=False, reason="Failed to Deserialize Final Response")
+        else:
+            return final_response
+
+    async def upload_dataset_item(self, dataset_name: str, item_name: str, source: Path) -> DatasetManagementResponse:
```

Comment:
Akin to my other comments, we should consider using a `Reader` as a source instead of a `Path` object.

Reply:
This should fall under #432 again.